src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,797 @@
+      Tensor<cpu, 1, DType> lvalue = data_l[iter_l++];
+      MXNET_ASSIGN_REQ_SWITCH(req, Req, {
+        Kernel<BinaryOpMissingRValue<OP, Req>, cpu>::Launch(
+          s, lvalue.shape_.Size(), out[iter_out++].dptr_, lvalue.dptr_);
+      });
+    }
+    while (iter_r < num_rows_r && !rhs_is_dense && !lhs_in_place) {
+      if (!is_dense_result) {
         indices_out[iter_out] = indices_r[iter_r];
-        Copy(out[iter_out++], data_r[iter_r++], s);
+      } else {
+        const IType idx_r = indices_r[iter_r];
+        iter_out = FillDense<cpu, DType, OP>(s, lhs.shape()[0], idx_r, req, &out, iter_out);
       }
-      auto new_sshape = TShape(output.aux_shape(rowsparse::kIdx));
-      CHECK_GT(new_sshape[0], num_common_rows);
-      new_sshape[0] -= num_common_rows;
-      output.set_aux_shape(rowsparse::kIdx, new_sshape);
-    });
-  });
-}
-
-template<typename xpu, typename OP>
-void BinaryComputeEx(const nnvm::NodeAttrs& attrs,
-                         const OpContext& ctx,
-                         const std::vector<NDArray>& inputs,
-                         const std::vector<OpReqType>& req,
-                         const std::vector<NDArray>& outputs) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  CHECK_EQ(inputs.size(), 2);
-  CHECK_EQ(outputs.size(), 1);
-  if (typeid(OP) == typeid(mshadow::op::plus)) {
-    // If any input is dense, fallback to FCompute
-    // TODO(haibin) implement dns + rsp in a separate kernel
-    if (common::ContainsDefaultStorage(inputs)) {
-      FCompExFallback<xpu>(attrs, ctx, inputs, req, outputs,
-                           BinaryCompute<xpu, OP>, "BinaryCompute");
+      Tensor<cpu, 1, DType> rvalue = data_r[iter_r++];
+      MXNET_ASSIGN_REQ_SWITCH(req, Req, {
+        Kernel<BinaryOpMissingLValue<OP, Req>, cpu>::Launch(
+          s, rvalue.shape_.Size(), out[iter_out++].dptr_, rvalue.dptr_);
+      });
+    }
+    if (is_dense_result) {
+      const size_t all_rows = static_cast<size_t>(lhs.shape()[0]);
+      iter_out = FillDense<cpu, DType, OP>(s, all_rows, all_rows, req, &out, iter_out);
+    } else {
+      if (lhs_in_place) {
+        CHECK_LE(iter_out, num_rows_l);
+      }
+      if (rhs_in_place) {
+        CHECK_LE(iter_out, num_rows_r);
+      }
+      DCHECK_LE(iter_out, num_rows_l + num_rows_r);  // Make sure that we didn't overrun
+      nnvm::TShape new_shape = output.aux_shape(rowsparse::kIdx);
+      CHECK_LE(iter_out, new_shape.Size());
+      if (!rhs_is_dense && !lhs_is_dense && !lhs_in_place && !rhs_in_place) {
+        // Reduce the first-dimension size by the number of common rows
+        new_shape[0] -= num_common_rows;
+        output.set_aux_shape(rowsparse::kIdx, new_shape);
+      }
+    }
+  }
+
+  /*! \brief CSR -op- CSR binary operator for non-canonical NDARray */
+  template<typename DType, typename IType, typename CType, typename OP>
+  static inline void CsrCsrElemwiseBinaryOp(mshadow::Stream<cpu> *s,
```

Comment:
This is function tested? I didn't notice binary csr from the operator list you made.